packages/teleport/src/services/websession/storeWebSession.ts

```diff
@@ -0,0 +1,163 @@
+    const time = Number(window.sessionStorage.getItem(KeysEnum.LAST_ACTIVE));
+    return time ? time : 0;
+  }
+
+  getAccessToken() {
+    const bearerToken = getBearerToken();
+    return bearerToken ? bearerToken.accessToken : null;
+  }
+
+  getSessionInactivityTimeout() {
+    const bearerToken = getBearerToken();
+    const time = Number(bearerToken.sessionInactiveTimeout);
+    return time ? time : 0;
+  }
+
+  initBroadcastChannel() {
+    this.bcBroadcaster = new BroadcastChannel('websession_store');
+    this.bcReceiver = new BroadcastChannel('websession_store');
+
+    this.bcReceiver.onmessage = ({
+      data,
+    }: MessageEvent<BroadcastChannelMessage>) => {
+      // UPSERT_TOKEN indicates that a token has been upserted from a tab, this
+      // causes all tabs to update their own sessionStorage with that token.
+      if (data.key === EventKeys.UPSERT_TOKEN) {
+        window.sessionStorage.setItem(
+          KeysEnum.TOKEN,
+          JSON.stringify(data.token)
+        );
+      }
+
+      // GET_TOKEN indicates that a tab is asking for the token, if a tab
+      // has a token in its sessionStorage, it upserts it so that other tabs
+      // receive it and update their own sessionStorage with it.
+      if (data.key === EventKeys.GET_TOKEN) {
+        const token = getBearerToken();
+        if (token) {
+          this.bcBroadcaster.postMessage({
+            key: EventKeys.UPSERT_TOKEN,
+            token,
+          });
+        }
+      }
+
+      if (data.key === EventKeys.UPSERT_TOKEN_IS_RENEWING) {
+        window.sessionStorage.setItem(
+          KeysEnum.TOKEN_RENEW,
+          data.isTokenRenewing.toString()
+        );
+      }
+
+      if (data.key === EventKeys.GET_TOKEN_IS_RENEWING) {
+        const isRenewing = this.getIsRenewing();
+        if (isRenewing) {
+          this.bcBroadcaster.postMessage({
+            key: EventKeys.UPSERT_TOKEN_IS_RENEWING,
+            isRenewing,
+          });
+        }
+      }
+
+      if (data.key === EventKeys.UPSERT_LAST_ACTIVE) {
+        window.sessionStorage.setItem(
+          KeysEnum.LAST_ACTIVE,
+          data.lastActive.toString()
+        );
+      }
+
+      if (data.key === EventKeys.GET_LAST_ACTIVE) {
+        const lastActive = this.getLastActive();
+        if (lastActive) {
+          this.bcBroadcaster.postMessage({
+            key: EventKeys.UPSERT_LAST_ACTIVE,
+            lastActive,
+          });
+        }
+      }
+
+      if (data.key === EventKeys.CLEAR) {
+        window.sessionStorage.clear();
+      }
+    };
+
+    if (!isTest) {
+      this.bcBroadcaster.postMessage({
+        key: EventKeys.GET_TOKEN,
+      });
+      this.bcBroadcaster.postMessage({
+        key: EventKeys.GET_TOKEN_IS_RENEWING,
+      });
+      this.bcBroadcaster.postMessage({
+        key: EventKeys.GET_LAST_ACTIVE,
+      });
+    }
+  }
+}
+
+export function getBearerToken(): BearerToken {
+  return JSON.parse(window.sessionStorage.getItem(KeysEnum.TOKEN));
```

Comment:
As this is grabbing data from the user space this can throw and because TS doesn't give you a way to indicate that something throws it's best to handle such an eventuality here.